### DISCUSSION TITLE
LookupError: url with Anvil Extras Routing on Startup After Returning from Vacation

### DISCUSSION THREAD

**Discussion by chad63e** [anvilistas/anvil-extras#519] [Q&A]
Hello,

I've encountered a `LookupError: url` with an Anvil app that was functioning correctly before I went on vacation. The app leverages the Anvil Extras routing library to manage URL routing based on the hash URL, primarily during the startup process. Although the app is still under development, and I'm currently focusing on testing specific pages rather than the full routing logic, this error has started to appear unexpectedly.

### Context:
- The app uses a startup module to parse the hash URL and route to the appropriate page.
- The setup includes imports for various Anvil services and the Anvil Extras routing library.
- I'm specifically testing the "contact-us" page at this moment.

### Startup Module Code Snippet:
```python
# Import statements
from anvil_extras import routing

from . import shared_import
from .globals import GLOBALS

def catch_hash():
    url_hash = anvil.get_url_hash()
    if GLOBALS.IN_DEBUG:
        routing.set_url_hash(
            "contact-us", url_dict=None, replace_current_url=True
        )
    routing.launch()

if __name__ == "__main__":
    catch_hash()
```

### Error Message:
```
LookupError: url
at app/anvil_extras/routing/_navigation.py:18
called from app/anvil_extras/routing/__init__.py:13
called from Forms.UserManagement.ContactUs, line 10
called from shared_import, line 11
called from startup, line 12
```

### Attempts to Resolve:
- Ensured all form imports are handled in `shared_imports`.
- Tried adjusting the import order, specifically placing the routing import after `shared_imports`.

Despite these attempts, the error persists. I'm seeking guidance on troubleshooting this issue further, particularly any insights into the `LookupError: url` with the Anvil Extras routing library in this context. Any advice or suggestions would be greatly appreciated.

Thank you!

**Comment by chad63e** [comment]
Further investigation into the `LookupError: url` issue revealed that the `history.state` object, critical for routing, lacks the expected `url` key, presenting instead with an entry like `{'idx': 0}`. This discrepancy seems to be why `_navigation.py` struggles to obtain the correct data for session setup.

In an attempt to bypass this problem, I manually updated the `history.state` in my app to include a `url` key with a value. This workaround proved effective, allowing the app to function as intended, which confirmed my suspicion about the missing `url` in `history.state` being the root cause of the error.

However, manually adjusting `history.state` is clearly not a viable long-term solution. I am now looking for insights into why `history.state` isn't populated as expected within the Anvil Extras routing context and for recommendations on managing `history.state` appropriately without resorting to manual updates.

Would greatly appreciate any advice, insights, or best practices on addressing this issue fundamentally. Thank you for your support!

## Update startup module code:
```python
from anvil import *
from anvil.js.window import history, location

# Update history.state before the first import of anvil_extras.routing
if history.state is not None or "url" not in history.state:
    state = {"url": location.hash, "pos": 0}
    history.replaceState(state, "", state["url"])

from anvil_extras import routing

from . import shared_import
from .globals import GLOBALS


def catch_hash():
    url_hash = anvil.get_url_hash()
    if GLOBALS.IN_DEBUG:
        routing.set_url_hash(
            url_pattern="contact-us",
            url_dict=None,
            replace_current_url=True,
        )

    routing.launch()


if __name__ == "__main__":
    catch_hash()

```

**Reply by s-cork** [reply]
thanks for that - we'll get that fixed

**Reply by s-cork** [reply]
(note i suspect that something in your browser is overriding the history state, since there has been no changes to this part of the code for years. Perhaps a browser extension is interfering. Try running the app in incognito)

Either way your suggested code fix will be merged soon to ensure the history state is the right shape.

**Comment by chad63e** [comment]
Stu, thank you for your timely response. Following your suggestion, I ran the app in incognito mode and observed a slightly different behavior as anticipated. Specifically, `dict(history.state)` returned as an empty dictionary, unlike the previous output (`{'idx': 0}`). Consequently, my app initialized successfully without needing the extra condition required for non-incognito sessions.

This variance in behavior between standard and incognito modes is intriguing and seems to hint at underlying issues with how `history.state` is managed or initialized. Your insights have been invaluable in identifying this discrepancy.

Please keep me updated on any progress or patches related to this issue. I'm eager to streamline my code by removing the workaround once a permanent fix is in place. Again, thank you for your assistance and looking forward to any updates.

**Reply by s-cork** [reply]
Yeah this suggests it could be one of your browser extensions messing with the history state.

**Comment by Tiepies** [comment]
This error suddenly popped up today in all our Anvil applications that use `from anvil_extras import routing` (regardless of what browser we were using or whether we used incognito mode or not). We've fixed it by upgrading to v2.6.2 (we were on v2.5.3), but were a bit surprised by this error just appearing in a locked version of anvil-extras.

I did notice that this error started occurring right before v2.7.0 was released. Is there any way that the occurrence of this error has anything to do with this version bump? I wouldn't know how, but it just seemed like an intriguing coincidence.
